File: linkedlist.rb
require 'pry'

class Node
  attr_accessor :val, :next

  def initialize(val, next_node)
    @val = val
    @next = next_node
  end
end

class LinkedList
  def initialize(val)
    @head = Node.new(val, nil)
  end

  def add(val)
    current = @head
    while current.next != nil
      current = current.next
    end
    current.next = Node.new(val, nil)
  end

  def delete(val)
    current = @head
    if current.val == val
      @head = current.next
      current
    else
      while current && current.next
        if current.next.val == val
          if current.next.next
            current.next = current.next.next
          else
            current.next = nil
          end
        end
        current = current.next
      end
    end

  end

  def return_list
  end
end

binding.pry

File: linkedlist.js
class Node {
  constructor(val, next = null) {
    this.val = val;
    this.next = next;
  }
}

class LinkedList {
  constructor(val) {
    this.head = new Node(val, null);
  }

  add(val) {
    let current = this.head;
    while (current && current.next) {
      current = current.next;
    }
    current.next = new Node(val);
  }

// delete 1 matching node
  delete(val) {
    let current = this.head;
    if (current.val === val) {
      this.head = current.next;
    } else {
      while (current && current.next) {
        if (current.next.val === val) {
          if (current.next.next) {
            current.next = current.next.next;
            return;
          } else {
            current.next = null;
            return;
          }
        }
        current = current.next;
      }
    }
  }

  return_list() {
    let current = this.head;
    let arr = [current];
    while (current.next) {
      current = current.next;
      arr.push(current);
    }
    return arr;
  }
}
